sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -188,14 +188,17 @@ static inline std::string GetHTTPMessagePreBody(Azure::Core::Http::Request const
 
 static void CleanupThread()
 {
+  // NOTE: Avoid using Log::Write in here as it may fail on MacOS,
+  // see issue: https://github.com/Azure/azure-sdk-for-cpp/issues/3224
+  // This method can wake up in de-attached mode after the application has been terminated.
+  // If that happens, trying to use `Log` would cause `abort` as it was previously deallocated.
```

Comment:
Wow... hmm 🤔
Interesting issue and catch.

I am guessing this applies to any other globals or statics as well that might be initialized on start up?

Reply:
yes, we need to avoid static objects depending on each other if we don't control the exact moment of their interaction. 

For instance, an A and B static components can depend on each other, B can call  A.Foo() , or A can call B.Foo(). But only if we know that its interaction is always guaranteed to happen during application lifetime.

Reply:
@CaseyCarter with this whole statics and initialization subtleties with C++, is there some viable or clean solution?

Reply:
It's generally impossible to have detached threads _and_ clean shutdown. See [[basic.start.term]/6](https://eel.is/c++draft/basic.start.term#6):

> -6- If there is a use of a standard library object or function not permitted within signal handlers ([[support.runtime]](https://eel.is/c++draft/support.runtime)) that does not happen before ([[intro.multithread]] (https://eel.is/c++draft/intro.multithread)) completion of destruction of objects with static storage duration and execution of `std​::​atexit` registered functions ([[support.start.term]])(https://eel.is/c++draft/support.start.term), the program has undefined behavior.
> [Note 3: If there is a use of an object with static storage duration that does not happen before the object's destruction, the program has undefined behavior. Terminating every thread before a call to `std​::​exit` or the exit from `main` is sufficient, but not necessary, to satisfy these requirements. These requirements permit thread managers as static-storage-duration objects. — end note]

This paragraph is hinting at the fact that after destroying the program's static storage duration objects `atexit` may tear down the standard library as well, at which time any interaction with the standard library (freeing memory is the one that surprises folks the most) is UB.

Reply:
> It's generally impossible to have detached threads

The way I made it work for our SDK if to make the detached thread to be a member of a class and to make a global static instance of that class.  Then, the thread must only access things within the same class.

Hopefully we never get the issue where the std lib is gone before our global class instance. :)

Reply:
> > It's generally impossible to have detached threads
> 
> The way I made it work for our SDK if to make the detached thread to be a member of a class and to make a global static instance of that class. Then, the thread must only access things within the same class.

(Actually looks at `curl.cpp`) Aha! This thread is never detached - we `join` it in `~CurlConnectionPool` (and less critically in `MoveConnectionBackToPool`). I should have remembered this. So the only issue here is ensuring that facilities it uses are torn down after the connection pool.